README.md
# yaqd-new-era

[![PyPI](https://img.shields.io/pypi/v/yaqd-new-era)](https://pypi.org/project/yaqd-new-era)
[![Conda](https://img.shields.io/conda/vn/conda-forge/yaqd-new-era)](https://anaconda.org/conda-forge/yaqd-new-era)
[![yaq](https://img.shields.io/badge/framework-yaq-orange)](https://yaq.fyi/)
[![black](https://img.shields.io/badge/code--style-black-black)](https://black.readthedocs.io/)
[![ver](https://img.shields.io/badge/calver-YYYY.M.MICRO-blue)](https://calver.org/)
[![log](https://img.shields.io/badge/change-log-informational)](https://gitlab.com/yaq/yaqd-new-era/-/blob/main/CHANGELOG.md)

yaq daemons for new era pump systems hardware

This package contains the following daemon(s):

- https://yaq.fyi/daemons/new-era-ne1000
- https://yaq.fyi/daemons/new-era-continuous-nextgen

CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/).

## [Unreleased]

### Changed
- rerendered avpr corresponding to yaq-traits 2022.3.0

## [2022.1.0]

### Added
- new daemon for dual pump systems using x2 firmware ("nextgen")

## [2021.10.0]

### Changed
- rerender avprs for recent properties in traits

### Added
- document prescence on conda-forge

## [2021.5.0]

### Changed
- avro transition

## [2020.05.0]

### Added
- initial release

[Unreleased]: https://gitlab.com/yaq/yaqd-new-era/-/compare/v2022.1.0...main
[2022.1.0]: https://gitlab.com/yaq/yaqd-new-era/-/compare/v2021.10.0...v2022.1.0
[2021.10.0]: https://gitlab.com/yaq/yaqd-new-era/-/compare/v2021.5.0...v2021.10.0
[2021.5.0]: https://gitlab.com/yaq/yaqd-new-era/-/compare/v2020.05.0...v2021.5.0
[2020.05.0]: https://gitlab.com/yaq/yaqd-new-era/-/tags/v2020.05.0


yaqd_new_era/_ne1000.py
__all__ = ["NE1000"]

import asyncio
from typing import Dict, Any
import serial  # type: ignore

from yaqd_core import ContinuousHardware

from .__version__ import __branch__


class NE1000(ContinuousHardware):
    _kind = "ne1000"
    _version = "0.1.0" + f"+{__branch__}" if __branch__ else ""
    traits = ["uses-uart", "uses-serial"]
    defaults: Dict[str, Any] = {}  # of config
    defaults["baud_rate"] = 19200
    defaults["address"] = 0
    defaults["limits"] = [0, 1e3]
    defaults["units"] = "ML"
    defaults["rate_units"] = "MM"
    defaults["low_noise"] = True

    def __init__(self, name, config, config_filepath):
        super().__init__(name, config, config_filepath)
        self.serial_port = config["serial_port"]
        self.baud_rate = config["baud_rate"]
        self.config_limits = config["limits"]
        self.diameter = config["diameter"]
        self.address = config["address"]
        self.limits = config["limits"]
        self._units = config["units"]
        self._rate_units = config["rate_units"]
        self._purging = False
        self._previous_infused = 0
        self._previous_withdrawn = 0
        self.ser = serial.Serial(port=self.serial_port, baudrate=self.baud_rate, timeout=1)
        # get rate from pump
        prompt, alarm, data = self._write("RAT")
        self._rate = float(data[:-3])
        # low noise mode
        self._write(f"LN {int(config['low_noise'])}")

    def close(self):
        self.ser.close()

    def direct_serial_write(self, message):
        self.ser.write(message.encode())

    def get_state(self):
        state = super().get_state()
        state["rate"] = self._rate
        return state

    def prime(self):
        self._busy = True
        self._loop.create_task(self._prime())

    async def _prime(self):
        # prime
        self.ser.write(f"{self.address} DIR WDR\r\n".encode())
        self.ser.write(f"{self.address} PUR\r\n".encode())

    def purge(self):
        self._busy = True
        self._loop.create_task(self._purge())

    async def _purge(self):
        # purge
        self._purging = True
        self._write("DIR INF")
        self._write("PUR")

    def _set_position(self, position):
        amount_from_here = position - self._position
        if abs(amount_from_here) < 0.001:
            return
        if amount_from_here > 0:
            self._write("DIR WDR")
        elif amount_from_here < 0:
            self._write("DIR INF")
        self._write(f"VOL {self._units}")
        # amount needs special formatting
        if abs(amount_from_here) > 10000:
            raise Exception
        vol = str(abs(amount_from_here))[:5]
        self._write(f"VOL {vol}")  # TODO: truncate charachters
        self._write(f"RUN")

    def set_rate(self, rate):
        self._rate = rate
        self._write(f"RAT C {self._rate} {self._rate_units}")

    async def update_state(self):
        while True:
            # prompt
            prompt, alarm, out = self._write("DIS")
            if prompt == "I":  # infusing
                self._busy = True
            elif prompt == "W":  # withdrawing
                self._busy = True
            elif prompt == "S":  # stopped
                self._busy = False
            elif prompt == "P":  # paused
                self._busy = False
            elif prompt == "X":  # purging
                self._busy = True
            else:
                self._busy = True
            # purging: once done, position goes to 0
            if self._purging and alarm == "S":
                self._purging = False
                self._position = 0
            # get current position
            if not alarm:
                infused = float(out[out.find("I") + 1 : out.find("W")])
                withdrawn = float(out[out.find("W") + 1 : out.find("ML")])
                self._position -= infused - self._previous_infused
                self._position += withdrawn - self._previous_withdrawn
                self._previous_infused = infused
                self._previous_withdrawn = withdrawn
                # wait
                if self._busy:
                    await asyncio.sleep(0.1)
                else:
                    self._write(f"CLD INF")
                    self._previous_infused = 0
                    self._write(f"CLD WDR")
                    self._previous_withdrawn = 0
                    await asyncio.sleep(1)
            else:
                await asyncio.sleep(0)

    def _write(self, command):
        try:
            self.ser.flush()
            out = f"{self.address} {command}\r\n"
            self.direct_serial_write(out)
            _in = self.ser.read_until(b"\x03").decode().strip()
            address = int(_in[1:3])
            if _in[3] == "A":
                prompt = None
                alarm = _in[5]
                data = None
            else:
                prompt = _in[3]
                alarm = None
                data = _in[4:]
            return prompt, alarm, data
        except UnicodeDecodeError:
            return self._write(command)


yaqd_new_era/_serial.py
import asyncio
import serial
from yaqd_core import aserial


MAX_ADDRESSES = 10


class SerialDispatcher:
    def __init__(self, port, baudrate):
        self.port = aserial.ASerial(
            port,
            baudrate,
            stopbits=serial.STOPBITS_ONE,
            bytesize=serial.EIGHTBITS,
            parity=serial.PARITY_NONE,
            eol=b"\x03",
        )
        self.workers = [(None, None, None, None)] * MAX_ADDRESSES


class SerialDispatcher:
    def __init__(self, port, baudrate):
        self.port = aserial.ASerial(
            port,
            baudrate,
            stopbits=serial.STOPBITS_ONE,
            bytesize=serial.EIGHTBITS,
            parity=serial.PARITY_NONE,
            eol=b"\x03",
        )
        self.workers = [(None, None, None, None)] * MAX_ADDRESSES
        self.write_queue = asyncio.Queue()
        self.read_queue = asyncio.Queue()
        self.loop = asyncio.get_event_loop()
        self.tasks = [
            self.do_writes(),
            self.read_dispatch(),
        ]

    def write(self, data):
        self.write_queue.put_nowait(data)

    def do_writes(self):
        self.loop.create_task(self._async_do_writes())

    async def _async_do_writes(self):
        while True:
            data = await self.write_queue.get()
            self.port.write(data)
            await asyncio.sleep(0.25)
            # self.write_queue.task_done()

    def read_dispatch(self):
        self.loop.create_task(self._async_read_dispatch())

    async def _async_read_dispatch(self):
        while True:
            line = await self.port.areadline()
            response = line.decode().strip()
            await self.read_queue.put(response)
            try:
                address = int(response[1:3])
                if (address >= 0) and (address <= 9):

                    prompt = alarm = error = data = None
                    if response[3] == "A":  # alarm
                        alarm = response[4:-1]
                    else:

                        prompt = response[3]
                    if response[4] == "?":  # error
                        error = response[5:-1]
                        if error == "":
                            error = None
                    else:
                        data = response[4:-1]
                        if data == "":
                            data = None
                    self.workers[address] = prompt, alarm, error, data
            except:
                pass
            await asyncio.sleep(0.25)

    def flush(self):
        self.port.flush()

    def close(self):
        self.loop.create_task(self._close())

    async def _close(self):
        await self.write_queue.join()
        for worker in self.workers.values():
            await worker.join()
        for task in self.tasks:
            task.cancel()


yaqd_new_era/_serial_x2.py
import asyncio
import serial
from yaqd_core import aserial
import time


MAX_ADDRESSES = 10


class SerialDispatcher:
    def __init__(self, port, baudrate):
        self.port = aserial.ASerial(
            port,
            baudrate,
            stopbits=serial.STOPBITS_ONE,
            bytesize=serial.EIGHTBITS,
            parity=serial.PARITY_NONE,
            eol=b"\x03",
        )
        self.workers = [(None, None, None, None)] * MAX_ADDRESSES
        self.write_queue = asyncio.Queue()
        self.read_queue = asyncio.Queue()
        self.loop = asyncio.get_event_loop()
        self.tasks = [
            self.do_writes(),
            self.read_dispatch(),
        ]

    def write(self, data):
        self.write_queue.put_nowait(data)

    def do_writes(self):
        self.loop.create_task(self._async_do_writes())

    async def _async_do_writes(self):
        while self.port.is_open:
            if True:
                data = await self.write_queue.get()
                self.port.write(data)

    def read_dispatch(self):
        self.loop.create_task(self._async_read_dispatch())

    async def _async_read_dispatch(self):
        # This section is not fully configured for an X2 method.  Workers
        # fill an array equal to the total allowed addresses on a single
        # serial line (10).  If a dual syringe pump is installed, it does
        # not respond to an address and must instead be communicated to via
        # "*", but it returns an address of 0 and therefore fills that index.
        # But a "*" call will return values for all other pumps on the line
        # as well. The addressing/arraying method herein was felt to best
        # accomodate these varying serial setups.

        while self.port.is_open:
            if True:  # self.closing==False:
                line = await self.port.areadline()
                response = line.decode().strip()
                await self.read_queue.put(response)
                try:
                    address = int(response[1:3])
                    if (address >= 0) and (address <= 9):
                        prompt = alarm = error = data = None
                        if response[3] == "A":  # alarm
                            alarm = response[4:-1]
                        else:
                            prompt = response[3]
                        if response[4] == "?":  # error
                            error = response[5:-1]
                            if error == "":
                                error = None
                        else:
                            data = response[4:-1]
                            if data == "":
                                data = None
                        self.workers[address] = prompt, alarm, error, data
                except:
                    pass

    def flush(self):
        self.port.flush()

    def close(self):
        for task in self.tasks:
            task.cancel()  # NOTE: the task order may be important (cancel writes before reads)
            time.sleep(1.0)  # somewhere around what we would think a serial timeout to be
        self._close()
        self.port.close()

    def _close(self):
        pass


yaqd_new_era/_new_era_x2.py
import asyncio
import time
from typing import Dict, List, Optional

from yaqd_core import UsesUart, UsesSerial, IsHomeable, HasPosition, IsDaemon, IsDiscrete
from ._serial_x2 import SerialDispatcher


class NewEraX2(UsesUart, UsesSerial, IsHomeable, IsDiscrete, HasPosition, IsDaemon):
    _kind = "new-era-x2"

    errors = {
        "?": "Command unrecognized",
        "NA": "Command N/A",
        "OOR": "Command data out of range",
        "COM": "Invalid Serial packet",
        "IGN": "Command ignored",
    }

    prompts = {
        "I": "infusing",
        "W": "withdrawing",
        "S": "stopped",
        "P": "paused",
        "T": "timed pause",
        "U": "user wait",
        "X": "purging",
    }

    alarms = {
        "R": "Pump Reset (Power interrupted)",
        "S": "Pump Stall",
        "T": "Serial Timeout",
        "E": "Program error",
        "O": "Program phase number out of range",
        "U": "User Commanded Alarm (added)",
    }

    errors_dict = {value: key for key, value in errors.items()}
    alarms_dict = {value: key for key, value in alarms.items()}
    prompts_dict = {value: key for key, value in prompts.items()}

    serial_dispatchers: Dict[str, SerialDispatcher] = {}

    def __init__(self, name, config, config_filepath):
        super().__init__(name, config, config_filepath)
        self._homing = False
        self._busy = False
        addr = int(config["address"])
        if (addr > 9) or (addr < -1):
            return IndexError("pump systems currently only addressable in range 0-9")
        elif addr == -1:
            self._address = "*"
        else:
            self._address = str(addr)

        if self._address == "*":
            self._read_address = 0
        else:
            self._read_address = int(self._address)
        self.logger.info(f"address:{self._read_address}")
        if config["serial_port"] in self.serial_dispatchers:
            self._serial = self.serial_dispatchers[config["serial_port"]]
        else:
            self._serial = SerialDispatcher(config["serial_port"], baudrate=config["baud_rate"])
            self.serial_dispatchers[config["serial_port"]] = self._serial
        self._state["current_alarm"] = ""
        self._state["current_prompt"] = ""
        self.tasks = [
            self.read_from_serial(),
        ]

    def busy(self):
        return bool(super().busy())

    def _set_position(self, position):
        if float(position) != float(0.0):
            pos = True
        else:
            pos = False
        self._loop.create_task(self._wait_for_ready_and_set_position(pos))

    async def _wait_for_ready_and_set_position(self, position):
        if position:
            if ("P" in self._state["current_prompt"] or "S" in self._state["current_prompt"]) and (
                self._state["current_alarm"] == ""
            ):
                strn = f"{self._address}RUN\r"
                await self._serial.write_queue.put(strn.encode())
                self.logger.info(f"start: time {time.localtime()}")

            if self._busy and not self._homing:
                await self._not_busy_sig.wait()
                self._busy = True
        else:
            strn = f"{self._address}STP\r"
            await self._serial.write_queue.put(strn.encode())
            self.logger.info(f"stop: time {time.localtime()}")
            if self._busy and not self._homing:
                await self._not_busy_sig.wait()
                self._busy = True

    def set_alarm(self, alarm):
        assert isinstance(alarm, bool)
        self._loop.create_task(self._wait_for_ready_and_set_alarm(alarm))
        if alarm:
            self._state["current_alarm"] = "U"
            self.logger.info("type U alarm")
        else:
            self._state["current_alarm"] = ""

    async def _wait_for_ready_and_set_alarm(self, alarm):
        if alarm:
            self._busy = True
            strn1 = f"{self._address}BUZ1\r"
            strn2 = f"{self._address}OUT51\r"
            await self._serial.write_queue.put(strn1.encode())
            await self._serial.write_queue.put(strn2.encode())
            self.logger.info(f"alarm sounded  time:{time.localtime()}")
            self.set_position(0.0)
            if self._busy and not self._homing:
                await self._not_busy_sig.wait()
                self._busy = True
        else:
            self._busy = True
            strn1 = f"{self._address}OUT50\r"
            await self._serial.write_queue.put(strn1.encode())
            self.logger.info(f"alarm reset time:{time.localtime()}")
            if self._busy and not self._homing:
                await self._not_busy_sig.wait()
                self._busy = True
            self._state["current_alarm"] = ""

            
    def _set_position(self,position):

        async def _wait_for_ready_and_set_position(self, position):
            if float(position) != float(0.0):
                pos = True
            else:
                pos = False
            if pos:
                if (
                    "P" in self._state["current_prompt"] or "S" in self._state["current_prompt"]
                ) and (self._state["current_alarm"] == ""):
                    strn = f"{self._address}RUN\r"
                    await self._serial.write_queue.put(strn.encode())
                    self.logger.info(f"start: time {time.localtime()}")

                if self._busy and not self._homing:
                    await self._not_busy_sig.wait()
                    self._busy = True
            else:
                strn = f"{self._address}STP\r"
                await self._serial.write_queue.put(strn.encode())
                self.logger.info(f"stop: time {time.localtime()}")
                if self._busy and not self._homing:
                    await self._not_busy_sig.wait()
                    self._busy = True

        self._serial.loop.create_task(_wait_for_ready_and_set_position(self, position))

    def set_alarm(self, alarm):
        assert isinstance(alarm, bool)

        async def _wait_for_ready_and_set_alarm(self, alarm):
            if alarm:
                strn1 = f"{self._address}BUZ2\r"
                strn2 = f"{self._address}OUT51\r"
                await self._serial.write_queue.put(strn1.encode())
                await self._serial.write_queue.put(strn2.encode())
                al = self._state["current_alarm"]
                self.logger.info(f"alarm sounded  type:{al}  time:{time.localtime()}")

                if self._busy and not self._homing:
                    await self._not_busy_sig.wait()
                    self._busy = True
            else:
                strn1 = f"{self._address}OUT50\r"
                await self._serial.write_queue.put(strn1.encode())
                self.logger.info(f"alarm reset time:{time.localtime()}")
                if self._busy and not self._homing:
                    await self._not_busy_sig.wait()
                    self._busy = True

        self._serial.loop.create_task(_wait_for_ready_and_set_alarm(self, alarm))

    def get_alarm(self):
        al = self._state["current_alarm"]
        if al is not "":
            alarm = True
        else:
            alarm = False
        return alarm

    def read_from_serial(self):
        self._loop.create_task(self._read_from_serial())

    async def _read_from_serial(self):
        while self._loop.is_running():
            if True:  # held in place for a self.closing read if needed
                prompt, alarm, error, out = self._serial.workers[self._read_address]
                if prompt is not None:
                    self._state["current_prompt"] = prompt
                    if (prompt == "I") or (prompt == "W"):
                        self._state["position"] = 1.0
                        self._state[
                            "position_identifier"
                        ] = self.get_position_identifier_options()[1]
                    else:
                        self._state["position"] = 0.0
                        self._state[
                            "position_identifier"
                        ] = self.get_position_identifier_options()[0]
                    # self._state["current_alarm"]=""
                if alarm is not None:
                    if (alarm != self._state["current_alarm"]) and (
                        self._state["current_alarm"] != "U"
                    ):
                        self._serial.loop.create_task(self._wait_for_ready_and_set_alarm(True))
                        self._state["position"] = 0.0
                        self._state[
                            "position_identifier"
                        ] = self.get_position_identifier_options()[0]
                        self._state["current_alarm"] = alarm
                else:
                    alarm = ""
                    if (alarm != self._state["current_alarm"]) and (
                        self._state["current_alarm"] != "U"
                    ):
                        self.set_alarm(False)
                        self._state["current_alarm"] = ""
                        # log the alarm turn off?
                    pass
                if error is not None:
                    self.logger.info(f"command error: {error}")
                    pass
                if out is not None:
                    # self.logger.info(f"return: {out}")  #turn this back on if one needs to use direct_serial_write often
                    # otherwise, leave off as often the returned message is not well framed
                    pass
                await asyncio.sleep(0.25)

    def home(self):
        self._busy = True
        self._homing = True
        self._loop.create_task(self._home())

    async def _home(self):
        # await self.write_queue.put() info here
        self._homing = False
        self._busy = False
        pass

    def direct_serial_write(self, command: bytes):
        self._busy = True
        self._loop.create_task(self._direct_serial_write(command))

    async def _direct_serial_write(self, command: bytes):
        await self._serial.write_queue.put(command)

    def close(self):
        for task in self.tasks:
            task.cancel()
        self._loop.create_task(self._close())
        time.sleep(4.0)

    async def _close(self):
        while self._loop.is_running():
            await asyncio.sleep(0.25)
            self._loop.stop()
        self._serial.flush()
        self._serial.close()
        await asyncio.sleep(2.5)

    async def update_state(self):
        while True:
            if self._loop.is_running():
                strn = "*\r"
                await self._serial.write_queue.put(strn.encode())
                await asyncio.sleep(0.25)


if __name__ == "__main__":
    NewEraX2.main()


yaqd_new_era/_new_era_continuous_nextgen.py
import asyncio
import serial
import numpy as np
from ._new_era_x2 import NewEraX2

"""
The full command set, alarms, prompts, addressing (for multiple pumps) mechanisms,
and errors are found in the NE4000 manual.   Additional commands may be found for the X2
firmware.

Current daemon assumes only one pump system is in communication with the associated
serial ("COM") port.   To avoid errors in addressing, all commands are preceded by an
asterisk which bypasses the requirement for tagging an address to the serial command but
then does not allow other devices on the serial line.

When queried (*):
Prompts
I - infusing
W - withdrawing
S - stopped
P - paused
T - timed pause
U - user wait
X - purging

Alarms
R - Pump Reset (Power interrupted)
S - Pump Stall
T - Serial timeout
E - Program error
O - Program phase number out of range
U - User Commanded Alarm (added)

Errors (Prompt fields)
? - Command unrecognized
NA- Command N/A
OOR Command data out of range
COM Invalid Serial packet
IGN Command ignored
"""


class NewEraContinuousNextGen(NewEraX2):
    _kind = "new-era-continuous-nextgen"

    def __init__(self, name, config, config_filepath):
        super().__init__(name, config, config_filepath)
        self._rate = float(np.nan)
        self._rate_units = ""
        self._purging = False
        self.tasks.append(self._process_x2_data())
        self.get_rate()

    def get_rate(self):
        self._get_rate()
        return self._rate

    def get_rate_units(self):
        self._get_rate()
        return self._rate_units

    def _get_rate(self):
        async def _wait_for_ready_and_get_rate(self):
            strn = f"{self._address}RAT\r"
            await self._serial.write_queue.put(strn.encode())
            if self._busy and not self._homing:
                await self._not_busy_sig.wait()
                self._busy = True

        self._loop.create_task(_wait_for_ready_and_get_rate(self))

    def set_rate_units(self, units):
        assert isinstance(units, str)
        self.logger.info("rate units setter deactivated")
        # self._rate_units=units

    def set_rate(self, rate):
        async def _wait_for_ready_and_set_rate(self, rate):
            if self._state["current_alarm"] == "":
                rate = int(rate)
                await self._serial.write_queue.put(f"{self._address}RAT{rate}\n".encode())
                if self._busy and not self._homing:
                    await self._not_busy_sig.wait()
                    self._busy = True

        self._loop.create_task(_wait_for_ready_and_set_rate(self,rate))
    

    async def _process_x2_data(self):
        while True:
            prompt, alarm, error, data = self._serial.workers[self._read_address]
            # add data, alarm, error, prompt processing here and not in the parent...

            # I would have to set up a emitter for changes within self._serial.workers
            # if this processing step were to be streamlined...I did not want
            # to cross ports yet.  See serial for reason behind the array of workers
            def process_x2_rate(data):
                units=data[-2:]
                if ((units=="UM") or (units=="MM") or (units=="UH") or (units=="MH")):
                    self._rate=float(data[:-3])
                    self._rate_units=units

            if data is not None:
                process_x2_rate(data)
            await asyncio.sleep(0.25)

    #def process_x2_data(self):
    #    self._loop.create_task(self._process_x2_data())


yaqd_new_era/_new_era_ne1000.py
import re
import asyncio
import time
from typing import Dict, Any, List

from yaqd_core import IsDaemon, HasPosition, HasLimits, UsesSerial, UsesUart, aserial


dis_regex = re.compile(r"^I([\.\d]+)W([\.\d]+)")
rate_regex = re.compile(r"([\.\d]+)")

rate_units = {
    "microliter/minute": "UM",
    "milliliter/minute": "MM",
    "microliter/hour": "UH",
    "milliliter/hour": "MH",
}
volume_units = {"microliter": "UL", "milliliter": "ML"}


class NewEraNe1000(UsesUart, UsesSerial, HasLimits, HasPosition, IsDaemon):
    _kind = "new-era-ne1000"

    def __init__(self, name, config, config_filepath):
        super().__init__(name, config, config_filepath)
        self._ser = aserial.ASerial(
            self._config["serial_port"], baudrate=self._config["baud_rate"], eol="\x03".encode()
        )
        self._rate = None
        self._purging = False
        # the pump ignores the very first RS232 command it sees after power cycling for some reason
        # so we send DIS as a junk command to get things going
        self._ser.write(f"{self._config['address']} DIS\r".encode())
        self._loop.create_task(self._get_rate())  # cache the rate later

    def direct_serial_write(self, message: bytes):
        self._ser.write(message)

    def get_rate(self) -> float:
        return self._rate

    async def _get_rate(self):
        try:
            prompt, alarm, out = await self._write("RAT")
            match = rate_regex.match(out)
            self._rate = float(match[1])
        except TypeError:
            await self._get_rate()

    def prime(self):
        # withdraw fully
        self.logger.info("Priming.")
        self._busy = True
        self._ser.write(f"{self._config['address']} DIR WDR\r".encode())
        self._ser.write(f"{self._config['address']} PUR\r".encode())

    def purge(self):
        # purge fully
        # this also implies that the new position is zero
        self.logger.info("Purging.")
        self._busy = True
        self._purging = True
        self._ser.write(f"{self._config['address']} DIR INF\r".encode())
        self._ser.write(f"{self._config['address']} PUR\r".encode())

    def _set_position(self, position: float) -> None:
        self._loop.create_task(self._aset_position(position))

    async def _aset_position(self, position: float) -> None:
        await self._write("DIR STP")
        amount_from_here = position - self._state["position"]
        if abs(amount_from_here) > 99_999:
            raise ValueError("cannot represent a number that large over New Era RS232 interface!")
        elif abs(amount_from_here) < 0.001:
            return
        # set direction
        direction = "WDR" if amount_from_here > 0 else "INF"
        await self._write(f"DIR {direction}")
        # set volume
        units = volume_units[self._config["volume_units"]]
        await self._write(f"VOL {units}")
        vol = f"{abs(amount_from_here):0.3f}"[:5]
        await self._write(f"VOL {vol}")
        # run
        await self._write("RUN")

    def set_rate(self, rate):
        units = rate_units[self._config["rate_units"]]
        rate = f"{rate:0.3f}"[:5]
        self._ser.write(f"{self._config['address']} RAT C {rate} {units}\r".encode())
        self._loop.create_task(self._get_rate())

    async def update_state(self):
        while True:
            start_position = self._state["position"]
            while True:
                prompt, alarm, out = await self._write("DIS")
                if alarm is not None:
                    continue
                self.logger.debug(f"{prompt}, {alarm}, {out}")
                if not out.startswith("I"):
                    continue
                # prompt
                self._busy = prompt not in ("S", "P")
                # get current position
                try:
                    match = dis_regex.match(out)
                    infused = float(match[1])
                    withdrawn = float(match[2])
                    self._state["position"] = start_position - infused + withdrawn
                except (ValueError, TypeError) as e:
                    self.logger.error(e)
                    continue
                if not self._busy:
                    break
            # purging: once done, position goes to 0
            if self._purging:
                await self._write("STP")
                self._purging = False
                self._state["position"] = 0
                self._state["destination"] = 0
            # once busy is released, clear values from pump
            await self._write("CLD INF")
            await self._write("CLD WDR")
            try:
                await asyncio.wait_for(self._busy_sig.wait(), timeout=5.0)
            except asyncio.TimeoutError:
                pass

    async def _write(self, command):
        try:
            await asyncio.sleep(0.1)
            self._ser.reset_input_buffer()
            self._ser.flush()
            out = f"{self._config['address']} {command}\r".encode()
            response = await self._ser.awrite_then_readline(out)
            response = response.decode().strip()
            self.logger.debug(f"Recieved response: {response}")
            address = int(response[1:3])
            prompt = alarm = data = None
            if response[3] == "A":  # alarm
                alarm = response[5]
            else:
                prompt = response[3]
                data = response[4:]
            return prompt, alarm, data
        except (UnicodeDecodeError, ValueError) as e:  # try again
            self.logger.error(e)
            return await self._write(command)
